File: java/src/main/java/pl/adambanaszkiewicz/four_layers_example/application_layer/to_this/application/usecase/AddProductToCart.java
package pl.adambanaszkiewicz.four_layers_example.application_layer.to_this.application.usecase;

import pl.adambanaszkiewicz.four_layers_example.application_layer.to_this.domain.CartRepositoryInterface;
import pl.adambanaszkiewicz.four_layers_example.application_layer.to_this.domain.Cart;
import pl.adambanaszkiewicz.four_layers_example.application_layer.to_this.domain.ProductDTO;
import pl.adambanaszkiewicz.four_layers_example.application_layer.to_this.domain.DomainException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.web.bind.annotation.*;
import org.springframework.stereotype.Service;

@Service
public class AddProductToCart {
    private final CartRepositoryInterface repository;

    @Autowired
    public AddProductToCart(
        @Qualifier("ApplicationLayerToThisCartRepository") CartRepositoryInterface repository
    ) {
        this.repository = repository;
    }

    public void execute(
        String productId,
        int qty,
        String userId
    ) throws Exception {
        // Begin transaction
        try {
            ProductDTO product = getProduct(productId);

            Cart cart = repository.get(userId);
            cart.addProduct(product, qty);

            repository.save(cart);

            // Commit transaction
        } catch (DomainException e) {
            // Do some domain logic after domain exception
            throw e;
        } catch (Exception e) {
            // Rollback transaction
            throw e;
        }
    }

    private ProductDTO getProduct(String productId) {
        int maxProductsToBuy = 2;
        int stock = 10;

        /**
         * To simplify code, create DTO directly here. In real world we would
         * need to fetch product from database, and create ProductDTO instance.
         * ProductDTO is here as a bridge between data from source
         * (database or external module) and our Domain in this module.
         * We could use separate Service to fetch this DTO.
         */
        return new ProductDTO(productId, true, stock, maxProductsToBuy);
    }
}

File: php/src/UserInterface/FromThis/Application/UseCase/AddProductToCart.php
<?php

declare(strict_types=1);

namespace App\UserInterfaceLayer\FromThis\Application\UseCase;

use App\UserInterfaceLayer\FromThis\Domain\CartRepositoryInterface;
use App\UserInterfaceLayer\FromThis\Domain\ProductDTO;

final class AddProductToCart
{
    public function __construct(
        private readonly CartRepositoryInterface $repository,
    ) {
    }

    public function __invoke(
        string $productId,
        int $qty,
        string $userId,
    ): void {
        // Begin transaction
        try {
            $product = $this->getProduct($productId);

            $cart = $this->repository->get($userId);
            $cart->addProduct($product, $qty);

            $this->repository->save($cart);

            // Commit transaction
        } catch (\DomainException $e) {
            // Do some domain logic after domain exception
            throw $e;
        } catch (\Exception $e) {
            // Rollback transaction
            throw $e;
        }
    }

    private function getProduct(string $productId): ProductDTO
    {
        $maxProductsToBuy = 2;
        $stock = 10;

        /**
         * To simplify code, create DTO directly. In real world we would
         * need to fetch product from database, and create ProductDTO instance.
         * ProductDTO is here as a bridge between data from source
         * (database or external module) and our Domain in this module.
         */
        return new ProductDTO($productId, true, $stock, $maxProductsToBuy);
    }
}
